daq/combine_reports_from_date_range.py

```diff
@@ -0,0 +1,106 @@
+import os
+import logger 
+import sys
+import datetime
+import re
+import json
+import functools
+
+import configurator
+from gcp import GcpManager
+from report import MdTable
+
+LOGGER = logger.get_logger('combine_reports_from_date_range')
+
+def _render_results(results):
+    def reduce_labels(sets):
+        return sorted(functools.reduce(lambda cur, acc: acc.union(cur), sets, set())) 
+    result_label_sets = map(lambda test: set(test.keys()), results['tests'].values())
+    result_labels = reduce_labels(result_label_sets)
+    tests_table = MdTable(['test'] + result_labels)
+    if results['tests']:
+        for test, result in results['tests'].items():
+            tests_table.add_row([test, *[str(result.get(label, 0)) for label in result_labels]]) 
+    else:
+        tests_table.add_row(['No results'])
+
+    category_label_sets = map(lambda category: set(category.keys()), results['categories'].values())
+    result_labels = reduce_labels(category_label_sets)
+    categories_table = MdTable(['categories'] + result_labels)
+    if results['categories']:
+        for category, result in results['categories'].items():
+            categories_table.add_row([category, *[str(result.get(label, 0)) for label in result_labels]]) 
+    else:
+        categories_table.add_row(['No results'])
+    
+    missing_table = MdTable(['missing tests', 'count'])
+    if results['missing']:
+        for test in sorted(results['missing'].keys()):
+            missing_table.add_row([test, str(results['missing'][test])]) 
+    else:
+        missing_table.add_row(['None'])
+    
+    return "\n%s\n" * 3 % (tests_table.render(), categories_table.render(), missing_table.render())    
+
+def main(device, start=None, end=None, gcp=None, reports_dir=os.path.join('inst', 'reports')):
+    aggregate = {'tests': {}, 'categories': {}, 'missing': {}}
+    if gcp:
+        json_reports = gcp.get_reports_from_date_range(device, start=start, end=end)    
+    else: 
+        LOGGER.info('Looking for reports locally')
+        prog = re.compile('^report_%s_\d{4}-\d{2}-\d{2}T\d{6}\+\d{4}.*\.json$' % device) 
+        json_files = [f for f in os.listdir(reports_dir) if prog.match(f)]
+        def get_json_reports():
+            for json_file in json_files:  
+                timestamp = list(json_file.split('_')[-1][:22])
+                timestamp[13:13], timestamp[16:16], timestamp[22:22] = list(":" * 3)
```

Comment:
extract constants to named constants.... otherwise, I have no idea what "13" represents.

Reply:
or -- just use .replace?

Reply:
This is because the : in timestamps are removed from report file names. This is adding the : back in so it can be parsed to a date time properly. But I’m not sure why the : are removed from filenames in the first place.